vms/platformvm/txs/fee/calculator.go

```
@@ -4,85 +4,172 @@
+	err := tx.Visit(c.c)
+	return c.c.fee, err
+}
+
+func (c *Calculator) AddFeesFor(complexity fees.Dimensions) (uint64, error) {
+	return c.c.addFeesFor(complexity)
+}
+
+func (c *Calculator) RemoveFeesFor(unitsToRm fees.Dimensions) (uint64, error) {
+	return c.c.removeFeesFor(unitsToRm)
+}
 
-	// this is guaranteed to never return an error
-	_ = tx.Visit(tmp)
-	return tmp.fee
+func (c *Calculator) GetCumulatedComplexity() fees.Dimensions {
+	if c.c.feeManager != nil {
+		return c.c.feeManager.GetCumulatedComplexity()
+	}
+	return fees.Empty
 }
 
-// calculator is intentionally unexported and used through Calculator to provide
-// a more convenient API
 type calculator struct {
-	// Pre E-fork inputs
-	upgrades  upgrade.Config
+	// setup
+	isEActive bool
 	staticCfg StaticConfig
-	time      time.Time
+
+	// Pre E-upgrade inputs
+	upgrades upgrade.Config
+	time     time.Time
+
+	// Post E-upgrade inputs
+	feeManager         *fees.Manager
+	blockMaxComplexity fees.Dimensions
+	credentials        []verify.Verifiable
 
 	// outputs of visitor execution
 	fee uint64
 }
 
 func (c *calculator) AddValidatorTx(*txs.AddValidatorTx) error {
+	// AddValidatorTx is banned following Durango activation, so we
+	// only return the pre EUpgrade fee here
 	c.fee = c.staticCfg.AddPrimaryNetworkValidatorFee
 	return nil
 }
 
-func (c *calculator) AddSubnetValidatorTx(*txs.AddSubnetValidatorTx) error {
-	c.fee = c.staticCfg.AddSubnetValidatorFee
-	return nil
+func (c *calculator) AddSubnetValidatorTx(tx *txs.AddSubnetValidatorTx) error {
+	if !c.isEActive {
+		c.fee = c.staticCfg.AddSubnetValidatorFee
+		return nil
+	}
+
+	complexity, err := c.meterTx(tx, tx.Outs, tx.Ins)
+	if err != nil {
+		return err
+	}
+
+	_, err = c.addFeesFor(complexity)
+	return err
 }
 
 func (c *calculator) AddDelegatorTx(*txs.AddDelegatorTx) error {
+	// AddValidatorTx is banned following Durango activation, so we
+	// only return the pre EUpgrade fee here
```

Should we `err` here instead of silently returning the old fee?

---

Indeed it does allow us to simplify the constructor of the dynamic fee calculator